libmultilabel/metrics.py

```diff
@@ -31,68 +26,75 @@ class MultiLabelMetrics():
     def __init__(self, config):
         self.monitor_metrics = config.get('monitor_metrics', [])
         self.metric_threshold = config.get('metric_threshold', 0.5)
-        self.y_true = []
-        self.y_pred = []
-        self.cached_results = {}
+
+        self.n_eval = 0
+        self.multilabel_confusion_matrix = 0.
+        self.metric_stats = {}
 
         self.top_ks = set()
+        self.prec_recall_metrics = []
         for metric in self.monitor_metrics:
             if re.match('[P|R]@\d+', metric):
                 top_k = int(metric[2:])
                 self.top_ks.add(top_k)
+                self.metric_stats[metric] = 0.
+                self.prec_recall_metrics.append(metric)
             elif metric not in ['Micro-Precision', 'Micro-Recall', 'Micro-F1', 'Macro-F1', 'Another-Macro-F1']:
                 raise ValueError(f'Invalid metric: {metric}')
 
+    def reset(self):
+        self.n_eval = 0
+        self.multilabel_confusion_matrix = 0.
+        for metric in self.metric_stats:
+            self.metric_stats[metric] = 0.
+
     def add_values(self, y_true, y_pred):
         """Add batch of y_true and y_pred.
 
         Args:
             y_true (ndarray): an array with ground truth labels (shape: batch_size * number of classes)
-            y_pred (ndarray): an array with predicted labels (shape: batch_size * number of classes)
+            y_pred (ndarray): an array with predicted label values (shape: batch_size * number of classes)
         """
-        self.y_true.append(y_true)
-        self.y_pred.append(y_pred)
-
-    def eval(self):
-        """Evaluate precision, recall, micro-f1, macro-f1, and P@k/R@k listed in the monitor_metrics."""
-        y_true = np.vstack(self.y_true)
-        y_pred = np.vstack(self.y_pred)
-        report_dict = classification_report(y_true, y_pred > self.metric_threshold, output_dict=True, zero_division=0)
-        result = {
-            'Micro-Precision': report_dict['micro avg']['precision'],
-            'Micro-Recall': report_dict['micro avg']['recall'],
-            'Micro-F1': report_dict['micro avg']['f1-score'],
-            'Macro-F1': report_dict['macro avg']['f1-score'],
-            'Another-Macro-F1': another_macro_f1(
-                macro_prec=report_dict['macro avg']['precision'],
-                macro_rec=report_dict['macro avg']['recall'])  # caml's macro-f1
-        }
-        # add metrics like P@k, R@k to the result dict
```

Comment:
this comment should be kept and modified/expanded. As now we do a less clear thing of *n_eval